base/acme/src/org/dogtagpki/acme/server/ACMEEngine.java

```diff
@@ -222,4 +239,128 @@ public void contextDestroyed(ServletContextEvent event) {
             throw new RuntimeException("Unable to shutdown ACME engine: " + e.getMessage(), e);
         }
     }
+
+    public ACMENonce createNonce() throws Exception {
+
+        ACMENonce nonce = new ACMENonce();
+
+        // generate 128-bit nonce with JSS
+        // TODO: make it configurable
+
+        byte[] bytes = new byte[16];
+        SecureRandom random = SecureRandom.getInstance("pkcs11prng", "Mozilla-JSS");
+        random.nextBytes(bytes);
+        String value = Base64.encodeBase64URLSafeString(bytes);
+
+        nonce.setValue(value);
+
+        // set nonce to expire in 30 minutes
+        // TODO: make it configurable
+
+        long currentTime = System.currentTimeMillis();
+        long expirationTime = currentTime + 30 * 60 * 1000;
+
+        nonce.setExpirationTime(new Date(expirationTime));
+
+        database.addNonce(nonce);
+        logger.info("Nonce: " + nonce);
+
+        return nonce;
+    }
+
+    public void validateNonce(String value) throws Exception {
+
+        ACMENonce nonce = database.removeNonce(value);
+
+        if (nonce == null) {
+            throw new Exception("Invalid nonce: " + value);
+        }
+
+        long currentTime = System.currentTimeMillis();
+        long expirationTime = nonce.getExpirationTime().getTime();
+
+        if (expirationTime <= currentTime) {
+            throw new Exception("Expired nonce: " + value);
+        }
+
+        logger.info("Valid nonce: " + value);
+    }
+
+    public void purgeNonces() throws Exception {
+        database.removeExpiredNonces(new Date());
+    }
+
+    public void validateJWS(JWS jws, String alg, JWK jwk) throws Exception {
+
+        logger.info("Validating " + alg + " JWS");
+
+        String message = jws.getProtectedHeader() + "." + jws.getPayload();
+        byte[] signature = Base64.decodeBase64(jws.getSignature());
+
+        // JWS validation based on RFC 7515
+        // TODO: support other algorithms
+
+        if ("RS256".equals(alg)) {
+
+            String kty = jwk.getKty();
+            KeyFactory keyFactory = KeyFactory.getInstance(kty, "Mozilla-JSS");
+
+            String n = jwk.getN();
+            BigInteger modulus = new BigInteger(1, Base64.decodeBase64(n));
+
+            String e = jwk.getE();
+            BigInteger publicExponent = new BigInteger(1, Base64.decodeBase64(e));
```

Comment:
I don't know what NSS/JSS's keyFactory implementations validate, so I'll assume it'll happily take whatever we pass it and pretend it is a valid RSA key.

So we need a few safeguards here:

 - `e` needs to be sufficiently large. I'd disallow anything under `e=65537`.
 - `n` should be "sufficiently large" -- I'd make it be at least 2048 bits. 
 - We could argue whether or not we want to check a few small primes here. `2`, `3`, `5`, that type of thing. Ideally the underlying cryptographic library should check that for us so it'd be fast. Personally, I think better safer than sorry. But I've not yet checked what the impacts of being able to break this signature are.

Reply:
I don't think we should hard-code these kinds of crypto policies in PKI. I agree that the underlying crypto library should perform the checks.

Reply:
I think we need them somewhere, so I've filed an issue [against JSS](https://github.com/dogtagpki/jss/issues/298). These aren't "good to have" -- these are required checks for this signature to be meaningful at all. 

NSS will check them in FIPS mode, but otherwise doesn't validate keys at all.